[osdp_fixes 2/5] Various PD cleanups, enhancements and fixes

sidcha opened the pull request:
This PR brings in various updates such as bug fixes, enhancements and cleanups to subsys/mgmt/osdp in peripheral device mode. These changes have been tested with a st-nucleo-g474re board with secure channel enabled.

This PR is a split version of #46449 to improve review-ability (changes remain the same).

sidcha committed:
mgmt/osdp: Rename cmd_data as ephemeral_data

Since cmd_data member is used by both commands and events to store the
contents of current transaction, rename it to ephemeral_data which
better reflects the purpose of the variable.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Add capability checks on incoming commands

OSDP compliant devices communicate their capabilities and discover what
their peer can and cannot do. Right now, PD advertises these capabilities
and expects CP to honor them. Although this is not known to cause any
issues, it is not desirable to allow such accesses.

Add a check of incoming commands to to validate that the corresponding
capability was enabled and advertised.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Add support for key press and card read events

Now that we have the necessary infrastructure to collect events from PD
apps, we can use them to translate it to OSDP packet sequence for card
reads and key press events.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Rework MARK byte handling

OSDP specification section 5.7 states that a transmitting device has to
drive the transmission line to a marking state for a period of one char
in the current baud rate. This can be achieved by sending 0xFF. Since
this is not mentioned in the packet structure definition, many commercial
implementations of OSDP out in the wild do not send/expect this byte.

To work with such non-conforming devices, we will try to be as flexible
as possible in the PD: send mark byte only if the other side sent one. In
case of CP, we have no option but to send the mark byte to be as close
to the specification as possible. If a particular use case needs the CP
to not send it, we will provide a Kconfig option to disable it.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: phy: Catch out-of-order SC packets and fail

During handshake, only certain types of secure block types (<= SCS_14)
are allowed. A rouge CP/PD can try to bypass the handshake by directly
sending a secure block type ahead of the sequence and gain a secure
channel. Fix this by adding a check in packet decode time.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: Remove unused STR() macro

Initially, subsys/mgmt had its own STR() macro for string pasting which was
replaced with the zephry provided STRINGIFY(). The definition of this macro
seems to have lingered on so remove it.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Fix error reply code for CMD_KEYSET

By default, on errors, pd_decode_command replies with osdp_NAK with
sub-error code set to OSDP_PD_NAK_CMD_LEN (achieved using the ret ==
OSDP_PD_ERR_GENERIC check before return). This is works for all packet
framing errors; but when a more specific error code needs to be sent, ret
has to be set to something other than OSDP_PD_ERR_GENERIC (a suitable error
code happens to be OSDP_PD_ERR_REPLY) to prevent the tail check from
overwriting the error info.

In CMD_KEYSET, we fill a more specific error code but do not set the ret to
OSDP_PD_ERR_REPLY. It causes this error to be reported as a framing
error hence loosing some extended info about the error. Fix this issue by
reordering the checks a bit.

Fixes: 7f4d2c741b06 "mgmt/osdp: Add support for Secure Channel"
Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha committed:
mgmt/osdp: pd: Fix device capabilities report

Do not check or send the first entry in the pd->cp[] device capability
table which is for function code 0 which is not a defined function code.

Signed-off-by: David Vucich <dave@alcatraz.ai>
Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

gmarull commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -189,6 +189,69 @@ static bool do_command_callback(struct osdp_pd *pd, struct osdp_cmd *cmd)
+
+	pd->reply_id = REPLY_NAK;
+	pd->ephemeral_data[0] = OSDP_PD_NAK_CMD_UNKNOWN;
+	LOG_INF("PD is not capable of handling CMD(%02x); "
```
LOG_INF feels wrong here, maybe error/warn/dbg?

sidcha replied:
Yes, changing this to an error level log.

gmarull reviewed (dismissed):
lgtm, one minor nitpick

adakus commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -226,13 +292,21 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
 		if (len != CMD_ISTAT_DATA_LEN) {
 			break;
 		}
+		if (!pd_cmd_cap_ok(pd, NULL)) {
```
Are there any requirements defined for POLL command ?

sidcha replied:
No, all PD have to support the POLL command.

adakus replied:
So you can remove this capabilities check

sidcha replied:
This check is not in `case CMD_POLL`. It is for `case CMD_ISTAT`.

adakus commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -380,22 +466,26 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
 		if (len != CMD_KEYSET_DATA_LEN) {
 			break;
 		}
+		/* only key_type == 1 (SCBK) and key_len == 16 is supported */
```
Does spec defines any limitation for key length  ?

sidcha replied:
The spec defines the key size and block size to be 128 bits that give us the 16 bytes.

r2r0 replied:
In the 2.1.7 there is stated: "The Length indicates the number of bytes containing key data in the array Data[]. It is
computed as the integer value of the quantity of Key Length in bits plus 7 divided by 8. For
example, the Length shall be 16, and the Data[] array shall contain the 128-bit Secure
Channel base key (SCBK). For the 256-bit key the Length shall be 32."
Also in the 2.2 spec does not state exact 16 byte key length.

sidcha replied:
> In the 2.1.7 there is stated: "The Length indicates the number of bytes containing key data in the array Data[]. It is
computed as the integer value of the quantity of Key Length in bits plus 7 divided by 8. For
example, the Length shall be 16, and the Data[] array shall contain the 128-bit Secure
Channel base key (SCBK). 

Yes, I am aware of this. A lot of parts of mgmt/osdp is designed with the assumption `key_size = block_size = 16` (for that matter even the spec, see last point). If that were to change in a future version (right now there isn't any), then we need to take a look all over for such assumptions and fix them. For now, this check exists to assert it.

> For the 256-bit key the Length shall be 32.

I don't have 2.1.7, I think this has been removed in the latest version 2.2 of the spec (I cannot find it in 2.15 too).

> Also in the 2.2 spec does not state exact 16 byte key length

Not directly, no. But the spec make that obvious in various places. For instance,
- the SCBK default key is defined for 128 bits only.
- the capability mask defines only a 128 bit key.
- session key, MAC, derivation process talks about 128 bit keys only.

adakus commented on subsys/mgmt/osdp/src/osdp_phy.c:
```diff
@@ -131,16 +161,33 @@ int osdp_phy_packet_finalize(struct osdp_pd *pd, uint8_t *buf,
 
-	/* Do a sanity check only as we expect expect header to be prefilled */
-	if (buf[0] != OSDP_PKT_MARK || buf[1] != OSDP_PKT_SOM) {
+	/* Do a sanity check only; we expect expect header to be pre-filled */
```
expect expect

sidcha replied:
Ack, thanks.

r2r0 commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -635,6 +639,46 @@ static int pd_build_reply(struct osdp_pd *pd, uint8_t *buf, int max_len)
+		buf[len++] = BYTE_1(event->cardread.length);
+		for (i = 0; i < len_bytes; i++) {
+			buf[len++] = event->cardread.data[i];
+		}
```
Consider usnig `memcpy` for data copying.

sidcha replied:
Ack, thanks.

r2r0 commented on subsys/mgmt/osdp/src/osdp_pd.c:
```diff
@@ -269,6 +340,9 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
 		ret = OSDP_PD_ERR_REPLY;
+		if (!pd_cmd_cap_ok(pd, &cmd)) {
+			break;
+		}
```
Move capablity checking before setting any reply information.
It applies here and for CMD_LED, CMD_BUZ and CMD_TEXT.

sidcha replied:
Not sure I follow. Which reply information are you taking about?

r2r0 replied:
I.e. in this case `if (!pd_cmd_cap_ok(pd, &cmd)) {...}` should be moved just after `if (len != CMD_OUT_DATA_LEN || !pd->command_callback) {...}`

sidcha replied:
Ah, that. No we cannot do that; the cap check needs the command that we decoded in som cases (wherever cmd arg is non-null).

sidcha committed:
mgmt/osdp: Use memcpy instead of raw loops in many places

This patch replaces many instances where raw loops were used to copy bytes
with memcpy calls.

No functional change intended.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha commented:
@gmarull, Since your last review, 2 new (trivial) commits are added to this PR:
- 084841aa78 mgmt/osdp: Use memcpy instead of raw loops in many places
- 9ac2e5668d mgmt/osdp: pd: Fix device capabilities report

@adakus, @r2r0 I believe I have addressed all your comments

Please take another look.

r2r0 commented on subsys/mgmt/osdp/src/osdp_phy.c:
```diff
@@ -385,6 +385,12 @@ int osdp_phy_decode_packet(struct osdp_pd *pd, uint8_t *buf, int len)
 			return OSDP_ERR_PKT_FMT;
 		}
+		if (!sc_is_active(pd) && pkt->data[1] > SCS_14) {
+			LOG_ERR("Received invalid secure message!");
```
For CMD_SCRYPT there is LOG_WRN and here is LOG_ERR.
Both are cases where command is NACKed.

sidcha replied:
Should have been an error level log too. Fixed.

stephanosio commented:
@r2r0 @adakus It seems like you have not accepted the zephyrproject-rtos GitHub organisation invitation and your reviews are not counted towards the approvals required for merging this PR because of that (note that your approvals currently show up as grey). Please accept the invitation so that your reviews count and this PR can be merged.
